### PR TITLE
Implement `\#` to expand to all unshifted macro arguments

### LINKED ISSUE #596: [Feature request] An escape sequence that expands to all remaining macro arguments
This would have the same utility as `__VA_ARGS__` in C or `*args` in Python, allowing the remaining (un`SHIFT`ed) macro arguments to be passed on, possibly to another macro.

Candidates:

- `\0`: Not used for anything else; detected as a valid macro argument up until `expandMacroArg` explicitly fails it
- `\#`: It's a valid identifier character like `\@`; the number sign could be taken to mean "all numbers"
- `\*`: Star means "everything" in wildcards and regexes

Example usage:

```
items: MACRO
    db _NARG
    rept _NARG
        db \1
        shift
    endr
ENDM

; simplifies to :

items: MACRO
    db _NARG, \#
ENDM
````

```
; old way: person_event Y, X, (more than 10 more arguments, not necessarily a constant quantity)
; new way: object_event X, Y, (the rest of the arguments same as before)
; declare this for backwards compatibility
person_event: MACRO
Y = \1
X = \2
    shift 2
    ; no need to copy+paste the logic from object_event for the rest
    object_event X, Y, \#
ENDM
```

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #596

### CODE REVIEW

File: src/asm/macro.c
Comment (ISSOtm) [inline src/asm/macro.c:112]: Wouldn't it be better to return `NULL`, and have the caller not perform any expansion if that's returned?

File: src/asm/macro.c
Comment (ISSOtm) [inline src/asm/macro.c:119]: Make sure to check the returned value

File: src/asm/macro.c
Comment (ISSOtm) [inline src/asm/macro.c:121]: Arguably, no need to copy a terminator there

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/macro.c
Comment (Rangi42) [reply src/asm/macro.c:112]: `peek` calls `expandMacroArg` to "Expand the argument and return its first character". If no expansion is performed, `expandMacroArg` would have to return a sentinel `NULL`, and `peek` would have to continue on to the next character. Does that sound better? It has more special handling for the consequences of `\#` expanding to nothing, but avoids the arbitrary expansion to `" "`.

File: src/asm/macro.c
Comment (Rangi42) [reply src/asm/macro.c:121]: The terminator is necessary so the `strcat`s work.

Comment (ISSOtm) [reply src/asm/macro.c:124]: `strcat` re-reads the input string to find its terminator, when we already reached the end of the previous string (which we copied). I'd prefer maintaining a destination pointer in a variable, and incrementing it always.

File: src/asm/macro.c
Comment (ISSOtm) [reply src/asm/macro.c:112]: I think it'd be less overhead than allocating an expansion and then all the plumbing related to reading over it, so please go ahead with the former.

File: src/asm/macro.c
Comment (Rangi42) [reply src/asm/macro.c]: Done. It still reads each macro arg three times: once to get its length for calculating the total `\#` length; once to get its length for appending with `memcpy`; and once to actually do the `memcpy`. A custom loop to copy one character at a time, incrementing the `ptr` each time, would avoid that third read, but I expect `strlen`+`memcpy` is faster regardless.

Comment (ISSOtm) [reply src/asm/macro.c]: That would need to be benchmarked, but I'll accept this for the time being. We can always come back to performance tuning later.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:686]: [GCC doesn't like this, apparently](https://github.com/gbdev/rgbds/runs/1543959054?check_suite_focus=true#step:4:65). It should be fine to remove `const` from the struct, then.

Comment (Rangi42) [reply src/asm/lexer.c:686]: I was about to post a question about that. Removing the `const` from the struct only pushes the problem back, since various `const char *`s get assigned to `contents`. A lot of `const` qualifiers would need removing to avoid this warning.

Comment (Rangi42) [reply src/asm/lexer.c]: I think that's been solved by storing a `char *owned` instead of a `bool owned`, with it being `NULL` instead of `false` for typical un-owned const expansions.

Comment (Rangi42) [reply src/asm/lexer.c]: Now storing a `union` of `char *` and `const char *` instead. It's still a hacky workaround; `#pragma GCC diagnostic push/ignored/pop` might be preferable (it'd still be verbose but would keep the extra typing around the line where it's relevant).

File: src/asm/macro.c
Comment (Rangi42) [reply src/asm/macro.c]: Okay, done. This would also handle any other macro args which somehow expand to empty strings.

**Comment by ISSOtm** [comment]
[Checkpatch complains about trailing whitespace](https://github.com/gbdev/rgbds/pull/641/checks?check_run_id=1544129693#step:4:13) (I'm surprised your editor doesn't remove them?)

As for the `union`, maybe we ought to open an issue about warnings... let's keep that for now.

**Comment by Rangi42** [comment]
Fixed the `checkpatch` warning.

**Review by ISSOtm — APPROVED** [review]

**Comment by ISSOtm** [comment]
Rebased and fixed conflicts with PRs just merged; waiting for CI to greenlight, and merging.

**Pull request merged**
